contracts/test/v0.8/CronUpkeep.test.ts

```diff
@@ -0,0 +1,453 @@
+let cronTestHelper: CronInternalTestHelper
+let cronReceiver1: CronReceiver
+let cronReceiver2: CronReceiver
+
+let admin: SignerWithAddress
+let owner: SignerWithAddress
+let stranger: SignerWithAddress
+
+const timeStamp = 32503680000 // Jan 1, 3000 12:00AM
+
+let handler1Sig: string
+let handler2Sig: string
+let revertHandlerSig: string
+let basicSpec: string
+
+async function assertJobIDsEqual(expected: number[]) {
+  const ids = (await cron.getActiveCronJobIDs()).map((n) => n.toNumber())
+  assert.deepEqual(ids.sort(), expected.sort())
+}
+
+function decodePayload(payload: string) {
+  return utils.defaultAbiCoder.decode(
+    ['uint256', 'uint256', 'address', 'bytes'],
+    payload,
+  ) as [BigNumber, BigNumber, string, string]
+}
+
+function encodePayload(payload: [BigNumberish, BigNumberish, string, string]) {
+  return utils.defaultAbiCoder.encode(
+    ['uint256', 'uint256', 'address', 'bytes'],
+    payload,
+  )
+}
+
+async function createBasicCron() {
+  return await cron.createCronJobFromEncodedSpec(
+    cronReceiver1.address,
+    handler1Sig,
+    basicSpec,
+  )
+}
+
+describe('CronUpkeep', () => {
+  beforeEach(async () => {
+    const accounts = await ethers.getSigners()
+    admin = accounts[0]
+    owner = accounts[1]
+    stranger = accounts[2]
+    const crFactory = await ethers.getContractFactory('CronReceiver', owner)
+    cronReceiver1 = await crFactory.deploy()
+    cronReceiver2 = await crFactory.deploy()
+    const cronDelegateFactory = await ethers.getContractFactory(
+      'CronUpkeepDelegate',
+      admin,
+    )
+    const cronDelegate = await cronDelegateFactory.deploy()
+    const cronExternalFactory = await ethers.getContractFactory(
+      'src/v0.8/libraries/external/Cron.sol:Cron',
+      admin,
+    )
+    const cronExternalLib = await cronExternalFactory.deploy()
+    const cronFactory = await ethers.getContractFactory(
+      'CronUpkeepTestHelper',
+      {
+        signer: admin,
+        libraries: { Cron: cronExternalLib.address },
+      },
+    )
+    cron = (
+      await cronFactory.deploy(owner.address, cronDelegate.address)
+    ).connect(owner)
+    const fs = cronReceiver1.interface.functions
+    handler1Sig = utils.id(fs['handler1()'].format('sighash')).slice(0, 10) // TODO this seems like an ethers bug
+    handler2Sig = utils.id(fs['handler2()'].format('sighash')).slice(0, 10)
+    revertHandlerSig = utils
+      .id(fs['revertHandler()'].format('sighash'))
+      .slice(0, 10)
+    const cronTHFactory = await ethers.getContractFactory(
+      'CronInternalTestHelper',
+    )
+    cronTestHelper = await cronTHFactory.deploy()
+    basicSpec = await cron.cronStringToEncodedSpec('0 * * * *')
+  })
+
+  afterEach(async () => {
+    await h.reset()
+  })
+
+  it('has a limited public ABI', () => {
+    // Casting cron is necessary due to a tricky yarn version mismatch issue, likely between ethers
+    // and typechain. Remove once the version issue is resolved.
```

Comment:
can we make a new "bug" ticket for this and add the link here?

Reply:
done!

Reply:
nitpick, but it's nice to add a "TODO" at the beginning of the comment so devs know to try to fix this in the future